### ISSUE TITLE
Input image size of YOLOv8 is not 640

### ISSUE THREAD

**Issue by ruhuizhao** [ultralytics/ultralytics#4292]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, i have a question about the input size of YOLOv8, I notice that in the evaluation/metric, the input image shape is not 640? looks the model is using dynamic intput shape (depends on the image), like

```
at dataloader, img_i.shape  torch.Size([8, 3, 448, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 448, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 448, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 448, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 512, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 512, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 512, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 512, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 512, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 512, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 576, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 672, 672])
at dataloader, img_i.shape  torch.Size([8, 3, 672, 608])
at dataloader, img_i.shape  torch.Size([8, 3, 672, 512])
at dataloader, img_i.shape  torch.Size([8, 3, 672, 512])
at dataloader, img_i.shape  torch.Size([8, 3, 672, 448])
```

**Comment by glenn-jocher** [comment]
@ruhuizhao hello,

Thank you for using YOLOv8 and reaching out with your question. I'm happy to clarify your confusion around the input image size.

The image size inconsistency that you're observing doesn't necessarily mean that the model input shape is dynamic. Rather, it's a consequence of the dataset preprocessing steps, including image augmentation and resizing processes that aim to maintain the original aspect ratio of the images without distortion.

During these preprocessing steps, the longer side of the image is resized to the chosen size (for example, 640 pixels), while the other side is scaled in proportion. As a result, the final input size to the model may vary depending on the shape of the original image. This process helps to preserve the integrity of the features in images, which is important for successful detection. 

This is why you're seeing different image shape sizes in the log. The range of these sizes is conditioned by the aspect ratio of the images in your dataset.

I hope this clears up your question, and happy detecting using YOLOv8!

**Comment by ruhuizhao** [comment]
Thanks for your reply! but it's still not clear

> @ruhuizhao hello,
> 
> Thank you for using YOLOv8 and reaching out with your question. I'm happy to clarify your confusion around the input image size.
> 
> The image size inconsistency that you're observing doesn't necessarily mean that the model input shape is dynamic. Rather, it's a consequence of the dataset preprocessing steps, including image augmentation and resizing processes that aim to maintain the original aspect ratio of the images without distortion.
> 
I undertand that it's important to keep the original aspect ration for high quantily of objects. But the input shape of model is really strange. Yolov8 preprocessing looks differnet with yolov5, as yolov5 and v3 there is a padding step.
Below is the code for preprocessing during evaluation, we can see only image normalization but no reshape/transformations.
```
  def preprocess(self, batch):
      """Preprocesses batch of images for YOLO training."""
      batch['img'] = batch['img'].to(self.device, non_blocking=True)
      batch['img'] = (batch['img'].half() if self.args.half else batch['img'].float()) / 255
      for k in ['batch_idx', 'cls', 'bboxes']:
          batch[k] = batch[k].to(self.device)

      nb = len(batch['img'])
      self.lb = [torch.cat([batch['cls'], batch['bboxes']], dim=-1)[batch['batch_idx'] == i]
                 for i in range(nb)] if self.args.save_hybrid else []  # for autolabelling

```



> During these preprocessing steps, the longer side of the image is resized to the chosen size (for example, 640 pixels), while the other side is scaled in proportion. As a result, the final input size to the model may vary depending on the shape of the original image. This process helps to preserve the integrity of the features in images, which is important for successful detection.
> 
> This is why you're seeing different image shape sizes in the log. The range of these sizes is conditioned by the aspect ratio of the images in your dataset.
> 
This not the log, it's the print i add at engine/validator.py and yolo/detect/val.py preprocessing. They are exactlly the input image shape. 
I'm also puzzled by the args,  args.image_size==640 and this image_size is passed to the dataset.

![Screenshot 2023-08-10 at 20 27 42](https://github.com/ultralytics/ultralytics/assets/104890675/98623adf-780f-446f-8286-5d036d51aa55)

I'm using below code to generate the result, is it wrong? or The evaluation is indeed using dynamic image shape?

```
YOLOv8n = YOLO("yolov8n.pt")
data_cfg = "coco.yaml"
metrics = YOLOv8n.val(data=data_cfg, batch=8)
```
> I hope this clears up your question, and happy detecting using YOLOv8!

**Comment by glenn-jocher** [comment]
@ruhuizhao i see where the confusion is coming from and I apologize for any misunderstanding. In YOLOv8, there has been a change in the preprocessing steps compared to previous versions. Specifically, in YOLOv8, we don't have the padding step during the preprocessing.

The model indeed accepts varying image sizes. However, this doesn't mean that the model is "dynamic" per se. Rather, the model is designed to handle different aspect ratios more gracefully, leveraging the benefits from both fixed-size processing and keeping original aspect ratios.

The image preprocessing you are looking at, namely the normalization part, is just one aspect of the whole preprocessing pipeline. Image resizing actually happens before this step. When an image comes in, it gets resized so that its longer edge matches the specified size (e.g., 640 pixels), and the shorter edge gets resized proportionately. 

As for the dynamic anchor shapes, it's part of the design that enables YOLOv8 to handle different input sizes more efficiently. 

Your code snippet for evaluation should work fine. Based on your process, it seems like you're doing everything correctly. 

I hope this makes things clearer for you! Don't hesitate to ask if you have more questions.

**Comment by ruhuizhao** [comment]
I see, thanks for you explanation!

**Comment by glenn-jocher** [comment]
@ruhuizhao, you're welcome! I'm glad I could help clarify your questions. Don't hesitate to reach out if you have any more questions or need further explanation. Happy detecting with YOLOv8!
